libnettest2.hpp

```diff
@@ -0,0 +1,1805 @@
+        LIBNETTEST2_EMIT_WARNING("run: close_report() failed");
+        emit_ev("failure.report_close", {
+            {"failure", "library_error"},
+            {"library_error_context", err},
+        });
+      } else {
+        emit_ev("status.report_close", {{"report_id", ctx.report_id}});
+      }
+    }
+    emit_ev("status.progress", {{"percentage", 1.0},
+                                {"message", "report close"}});
+  } while (0);
+  // TODO(bassosimone): decide whether it makes sense to have an overall
+  // precise error code in this context (it seems not so easy). For now just
+  // always report success, which is what also legacy MK code does.
+  emit_ev("status.end", {{"failure", ""},
+                         {"downloaded_kb", info.bytes_down.load() / 1024.0},
+                         {"uploaded_kb", info.bytes_up.load() / 1024.0}});
+  return true;
+}
+
+void Runner::interrupt() noexcept { interrupted_ = true; }
+
+LogLevel Runner::get_log_level() const noexcept { return settings_.log_level; }
+
+// Methods you typically want to override
+// ``````````````````````````````````````
+
+void Runner::on_event(const nlohmann::json &event) const noexcept {
+  // When running with -fsanitize=thread enable on macOS, a data race in
+  // accessing std::clog is reported. Attempt to avoid that.
+  static std::mutex no_data_race;
+  std::unique_lock<std::mutex> _{no_data_race};
+  std::clog << event.dump() << std::endl;
+}
+
+// Methods you generally DON'T want to override
+// ````````````````````````````````````````````
+
+void Runner::emit_ev(std::string key, nlohmann::json value) const noexcept {
+  assert(value.is_object());
+  on_event({{"key", std::move(key)}, {"value", std::move(value)}});
+}
+
+bool Runner::run_with_index32(
+    const std::chrono::time_point<std::chrono::steady_clock> &begin,
+    const std::string &test_start_time,
+    const std::vector<std::string> &inputs, const NettestContext &ctx,
+    const std::string &collector_base_url, uint32_t i,
+    BytesInfo *info) const noexcept {
+  if (info == nullptr) return false;
+  {
+    auto current_time = std::chrono::steady_clock::now();
+    std::chrono::duration<double> elapsed = current_time - begin;
+    // We call a nettest done when we reach 90% of the expected runtime. This
+    // accounts for possible errors and for the time for closing the report.
+    if (settings_.max_runtime >= 0 &&
+        elapsed.count() >= settings_.max_runtime * 0.9) {
+      LIBNETTEST2_EMIT_INFO("exceeded max runtime");
+      return false;
+    }
+  }
+  emit_ev("status.measurement_start", {{"idx", i}, {"input", inputs[i]}});
+  nlohmann::json measurement;
+  measurement["annotations"] = settings_.annotations;
+  measurement["annotations"]["engine_name"] = settings_.engine_name;
+  measurement["annotations"]["engine_version"] = settings_.engine_version;
+  measurement["annotations"]["engine_version_full"] = settings_.engine_version;
```

Comment:
Why is this a duplicate of `engine_version`?

Reply:
This is most likely wrong. There should be a `engine_version_full` setting.

Reply:
See 6ae0241fdf88486baac4785b2502ad8db37e619c